File: ADT/djikstra's algo/2147.js
/**
 * @param {string} corridor
 * @return {number}
 */
var numberOfWays = function (corridor) {
    const MOD = 1000000007n;
    const seats = [];

    for (let i = 0; i < corridor.length; i++) {
        if (corridor[i] === 'S') {
            seats.push(i);
        }
    }

    // Invalid cases
    if (seats.length === 0 || seats.length % 2 !== 0) {
        return 0;
    }

    let ways = 1n;

    // Start from the 3rd seat, jump by pairs
    for (let i = 2; i < seats.length; i += 2) {
        const gap = BigInt(seats[i] - seats[i - 1]);
        ways = (ways * gap) % MOD;
    }

    return Number(ways);
};

File: ADT/djikstra's algo/2147.py
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        mod = int(1e9 + 7)
        pos_seats = [i for i, ch in enumerate(corridor) if ch == "S"]
        if len(pos_seats) % 2 != 0 or len(pos_seats) == 0:
            return 0
        result = 1
        prev = pos_seats[1]
        for i in range(2, len(pos_seats), 2):
            length = pos_seats[i] - prev
            result = (result * length) % mod
            prev = pos_seats[i] + 1
        return result
